src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -133,112 +145,11 @@ public final boolean isRunning() {
-                        startNodesAreClosed.put(((BlockEndNode) node).getStartNode().getId(), true);
-                    }
-                } // else we must have an old workflow-job
-            } // do not need to pay attention to atom nodes: either they are current heads, thus active, or they are not, thus inactive
-        }
-    }
-    @Restricted(DoNotUse.class)
-    @Extension public static final class FlowL extends FlowExecutionListener {
-        final Map<FlowExecutionOwner, Map<String, Boolean>> startNodesAreClosedByFlow = new HashMap<>();
-        static Map<FlowExecutionOwner, Map<String, Boolean>> startNodesAreClosedByFlow() {
-            FlowL flowL = ExtensionList.lookup(FlowExecutionListener.class).get(FlowL.class);
-            if (flowL == null) { // should not happen unless Jenkins is busted
-                throw new IllegalStateException("missing FlowNode.FlowL extension");
-            }
-            return flowL.startNodesAreClosedByFlow;
-        }
-        @Override public void onRunning(FlowExecution execution) {
-            LOGGER.finer("FlowExecutionListener working");
-            assert !startNodesAreClosedByFlow.containsKey(execution.getOwner());
-            startNodesAreClosedByFlow.put(execution.getOwner(), new HashMap<String, Boolean>());
-        }
-        @Override public void onResumed(FlowExecution execution) {
-            assert !startNodesAreClosedByFlow.containsKey(execution.getOwner());
-            Map<String, Boolean> startNodesAreClosed = new HashMap<String, Boolean>();
-            startNodesAreClosedByFlow.put(execution.getOwner(), startNodesAreClosed);
-            // To handle start nodes encountered in a prior Jenkins session, try to recreate the cache to date:
-            DepthFirstScanner dfs = new DepthFirstScanner();
-            dfs.setup(execution.getCurrentHeads());
-            for (FlowNode n : dfs) { // end nodes first, later the start nodes
-                if (n instanceof BlockEndNode) {
-                    startNodesAreClosed.put(((BlockEndNode) n).getStartNode().getId(), true);
-                } else if (n instanceof BlockStartNode) {
-                    if (!startNodesAreClosed.containsKey(n.getId())) {
-                        // If we have not encountered the BlockEndNode, it remains open.
-                        startNodesAreClosed.put(n.getId(), false);
-                    }
-                }
-            }
-        }
-        @Override public void onCompleted(FlowExecution execution) {
-            assert startNodesAreClosedByFlow.containsKey(execution.getOwner());
-            // After a build finishes, we do not need the cache any more, since we do the equivalent of FlowExecution.isComplete relatively quickly:
-            startNodesAreClosedByFlow.remove(execution.getOwner());
+        } if (this instanceof BlockEndNode || this instanceof AtomNode) {
+            return this.exec.isCurrentHead(this);
```

Comment:
I think you could just always delegate to `FlowExecution.isActive`, right? Seems like you are duplicating the special cases here.